eradiate/scenes/biosphere.py

```diff
@@ -586,125 +338,471 @@ def from_parameters(
+                                pos_candidate[1] + leaf_radius,
+                            ),
+                            (
+                                pos_candidate[2] - leaf_radius,
+                                pos_candidate[2] + leaf_radius,
+                            ),
+                        ]
+                    )
+                    if i == 0:
+                        positions[i] = pos_candidate
+                        tree.add(aabb)
+                        break
+                    else:
+                        if not tree.does_overlap(aabb):
+                            positions.append(pos_candidate)
+                            tree.add(aabb)
+                            break
+                else:
+                    raise RuntimeError(
+                        "unable to place all leaves: "
+                        "the specified canopy might be too dense"
+                    )
+    elif shape_type == "sphere":
+        if sphere_radius == 0:
+            raise ValueError(
+                "Parameter sphere_radius must be set for spherical leaf cloud."
+            )
+        for i in range(n_leaves):
+            rand = np.random.rand(3)
+            theta = rand[0] * np.pi
+            phi = rand[1] * 2 * np.pi
+            r = rand[2] * sphere_radius
+            positions[i] = [
+                r * np.sin(theta) * np.cos(phi),
+                r * np.sin(theta) * np.sin(phi),
+                r * np.cos(theta),
+            ]
+    elif shape_type == "cylinder":
+        if cylinder_radius == 0 or cylinder_height == 0:
+            raise ValueError(
+                "Parameters cylinder_height and cylinder_radius"
+                "must be set for cylindrical leaf cloud."
+            )
+
+        for i in range(n_leaves):
+            rand = np.random.rand(3)
+            phi = rand[0] * 2 * np.pi
+            r = rand[1] * cylinder_radius
+            z = rand[2] * cylinder_height
+            positions[i] = [r * np.cos(phi), r * np.sin(phi), z]
+
+    # compute leaf normals and create transformation matrices
+    orientations = np.empty((n_leaves, 3))
+    for i in range(np.shape(orientations)[0]):
+        theta = np.rad2deg(_inversebeta(mu, nu))
+        phi = np.random.rand() * 360.0
+
+        orientation = [
+            np.sin(theta) * np.cos(phi),
+            np.sin(theta) * np.sin(phi),
+            np.cos(theta),
+        ]
+        orientations[i] = orientation
+
+    return positions, orientations
+
+
+@attr.s
+class Canopy(SceneElement, ABC):
+    """
+    An abstract base class defining a base type for all canopies.
+    """
+    id = attr.ib(
+        default="canopy",
+        validator=attr.validators.optional(attr.validators.instance_of(str)),
+    )
+
+    size = attrib_quantity(
+        default=None,
+        validator=validator_has_len(3),
+        units_compatible=cdu.generator("length"),
+    )
+
+
+@attr.s
+class AbstractDiscreteCanopy(Canopy):
```

Comment:
How do you feel about renaming these classes and using the naming that is used in RAMI? 
For example the canopy we have already could just be called `HomogeneousCanopy` and in the long term it would have an interface, that lets users pick the discrete or the turbid implementation.

The annoying thing is this: If you look at the two canopies with floating spheres, there is one with spheres with FSOS interaction and one with floating disks inside the spheres. These two scenarios are called `HET01` and `HET02`. So I can't just create one class for these two and call it `HET01Canopy` because users will definitely wonder, where they will find the `HET02Canopy` then :man_facepalming: 

But a descriptive naming scheme like `FloatingSpheresCanopy` will also quickly get unusable, when we implement the two-component canopies, like the real-zoom-in.

Even among canopies where you would think this would be more straight forward, there are strange things:
The HomogeneousDiscreteCanopy we have right now, is called `HET13`. The purist corner scenario is `HET10`, `HET11` and `HET12`. There are text files which describe the leaves in these four canopies and they are actually all different. Yet it is obvious that there should not be individual classes for these four scenarios :man_facepalming: 

I guess we should go for the descriptive naming and work the names out on the fly.

Reply:
Good point: it's safer to use names of our making and have a good (_i.e._ as robust as possible) classification; we can then document the correspondance between RAMI test cases and our nomenclature, although I'm not even sure if we'll have to at all: if our test cases and components are well-written, things will should fairly obvious to the reader.